### PR TITLE
2771 multiprocess safe caching agent binary repository

### LINKED ISSUE #2771: Make `CachingAgentBinaryRepository` work with multiprocessing
# Refactor

## Component(s) to be refactored

- `infection_monkey.exploit.CachingAgentBinaryRepository`

## Explanation

The `CachingAgentBinaryRepository` currently uses the `lru_cache()` decorator to cache binaries downloaded from the island. Each plugin process will have its own copy of the cache, which means caching will not work properly.

Instead of using `lru_cache()`, cache the agent binaries in a shared dictionary. Use a `Manager` [[1](https://docs.python.org/3/library/multiprocessing.html#sharing-state-between-processes)][[2](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Manager)] to create a shared dictionary used for caching agent binaries. Don't forget to create the manager with a "spawn" context object (i.e. **do not** construct it with `multiprocessing.Manager()`)

- [ ] Cache agent binaries in a shared dictionary created with a `Manager` @cakekoa 
- [ ] Add a unit test that launches multiple processes that share the same `CachingAgentBinaryRepository` and verify only one call to the island is made for each `OperatingSystem`. @cakekoa

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Fixes #2771.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [ ] ~~Was the CHANGELOG.md updated to reflect the changes?~~
* [ ] ~~Was the documentation framework updated to reflect the changes?~~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running the unit tests
* [ ] ~~If applicable, add screenshots or log transcripts of the feature working~~

### CODE REVIEW

File: monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    p1 = context.Process(target=get_binary, args=(caching_agent_binary_repository, queue))
+    p1.start()
+    p1_binary = queue.get()
+    p1.join()
+
+    p2 = context.Process(target=get_binary, args=(caching_agent_binary_repository, queue))
+    p2.start()
+    p2_binary = queue.get()
+    p2.join()
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py:71]: Can we add another test that does this concurrently?

```
    p1 = context.Process(target=get_binary, args=(caching_agent_binary_repository, queue))
    p2 = context.Process(target=get_binary, args=(caching_agent_binary_repository, queue))

    p1.start()
    p2.start()

    p1_binary = queue.get()
    p2_binary = queue.get()

    p1.join()
    p2.join()
```

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        return self._calls.value
+
+    def get_agent_binary(self, _: OperatingSystem) -> bytes:
+        with self._calls.get_lock():
+            self._calls.value += 1
+            print(f"CALLED: {self._calls}")
+            return AGENT_BINARY
+
+
+def test_get_agent_binary__uses_cache(mock_island_api_client):
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py:39]: It's also worth adding a test that if I call get() for 2 different operating systems, I get different binaries back.

File: monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+
+@pytest.fixture
+def mock_island_api_client() -> IIslandAPIClient:
+    return StubIslandAPIClient()
+
+
+class PickableMagicMock(MagicMock):
+    def __reduce__(self):
+        return (MagicMock, ())
```
[the remark this answers is not in the document]
Comment (cakekoa) [reply monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py:21]: Thanks! I'll remove it

File: monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py (hunk 1 above)
Comment (cakekoa) [reply monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py:71]: Sure thing!

File: monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py (hunk 2 above)
Comment (cakekoa) [reply monkey/tests/unit_tests/infection_monkey/exploit/test_caching_agent_binary_repository.py:39]: No problemo!

**Comment by cakekoa** [comment]
These tests seem a bit slow -- should I mark them as such?

**Comment by mssalvatore** [comment]
> These tests seem a bit slow -- should I mark them as such?

Sounds like a good idea.

**Pull request merged**